File: backend/routers/project.py
"""
Project CRUD and node tree management router.
"""
from fastapi import APIRouter, HTTPException
from typing import List, Dict, Any
import uuid
from datetime import datetime
import logging

from shared.models import (
    Project, AgentNode, NodeStatus,
    CreateProjectRequest, AddNodeRequest,
)
from shared.state import get_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/projects", tags=["projects"])


def _get_project_mgr():
    mgr = get_service("project_manager")
    if mgr is None:
        raise HTTPException(status_code=500, detail="Project manager not initialized")
    return mgr


def _get_node_state():
    mgr = get_service("node_state")
    if mgr is None:
        raise HTTPException(status_code=500, detail="Node state manager not initialized")
    return mgr


@router.post("", response_model=Project)
async def create_project(req: CreateProjectRequest):
    """Create a new project with a local directory."""
    try:
        project_mgr = _get_project_mgr()
        project = project_mgr.create_project(req.name)
        return project
    except FileExistsError as e:
        raise HTTPException(status_code=409, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("", response_model=List[Project])
async def list_projects():
    """List all known projects."""
    project_mgr = _get_project_mgr()
    return project_mgr.list_projects()


@router.get("/{project_id}", response_model=Project)
async def get_project(project_id: str):
    """Get project details."""
    project_mgr = _get_project_mgr()
    project = project_mgr.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project


@router.delete("/{project_id}")
async def delete_project(project_id: str):
    """Delete a project."""
    project_mgr = _get_project_mgr()
    ok = project_mgr.delete_project(project_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Project not found")
    return {"status": "deleted"}


@router.post("/{project_id}/nodes", response_model=AgentNode)
async def add_node(project_id: str, req: AddNodeRequest):
    """Add an agent node to a project."""
    project_mgr = _get_project_mgr()
    node_state = _get_node_state()

    project = project_mgr.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    node = AgentNode(
        id=f"node-{uuid.uuid4().hex[:8]}",
        agent_name=req.agent_name,
        domain=req.domain,
        status=NodeStatus.PENDING,
        workflow_steps=req.workflow_steps or [],
        total_steps=len(req.workflow_steps or []),
        deliverables=[],
        context_files=req.context_files or [],
        created_at=datetime.utcnow().isoformat(),
    )

    result = project_mgr.add_node(project_id, node)
    if result is None:
        raise HTTPException(status_code=500, detail="Failed to add node")

    node_state.register(node.id, NodeStatus.PENDING)

    # Audit: node created
    audit_logger = get_service("audit_logger")
    if audit_logger:
        import asyncio
        from shared.models import AuditEventType
        asyncio.ensure_future(audit_logger.log_event(
            project_id=project_id,
            node_id=node.id,
            event_type=AuditEventType.NODE_CREATED,
            data={"agent_name": req.agent_name, "domain": req.domain},
            actor="user",
        ))

    return node


@router.put("/{project_id}/nodes/{node_id}", response_model=AgentNode)
async def update_node(project_id: str, node_id: str, updates: Dict[str, Any]):
    """Update a node's properties."""
    project_mgr = _get_project_mgr()

    project = project_mgr.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    updated = project_mgr.update_node(project_id, node_id, **updates)
    if updated is None:
        raise HTTPException(status_code=404, detail="Node not found")
    return updated


@router.delete("/{project_id}/nodes/{node_id}")
async def remove_node(project_id: str, node_id: str):
    """Remove a node from a project."""
    project_mgr = _get_project_mgr()
    node_state = _get_node_state()

    project = project_mgr.get_project(project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    ok = project_mgr.remove_node(project_id, node_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Node not found")

    node_state.remove(node_id)
    return {"status": "removed"}

File: client/src/main/ipc/project.js
const axios = require('axios')

function register(ipcMain, config) {
  const baseUrl = config.PYTHON_BACKEND_URL

  // Create a new project
  ipcMain.handle('project:create', async (_event, data) => {
    try {
      const response = await axios.post(`${baseUrl}/projects`, data)
      return { success: true, data: response.data }
    } catch (err) {
      return {
        success: false,
        error: err.response?.data?.detail || err.message
      }
    }
  })

  // List all projects
  ipcMain.handle('project:list', async () => {
    try {
      const response = await axios.get(`${baseUrl}/projects`)
      return { success: true, data: response.data }
    } catch (err) {
      return {
        success: false,
        error: err.response?.data?.detail || err.message
      }
    }
  })

  // Get a single project by ID
  ipcMain.handle('project:get', async (_event, projectId) => {
    try {
      const response = await axios.get(`${baseUrl}/projects/${projectId}`)
      return { success: true, data: response.data }
    } catch (err) {
      return {
        success: false,
        error: err.response?.data?.detail || err.message
      }
    }
  })

  // Delete a project
  ipcMain.handle('project:delete', async (_event, projectId) => {
    try {
      const response = await axios.delete(`${baseUrl}/projects/${projectId}`)
      return { success: true, data: response.data }
    } catch (err) {
      return {
        success: false,
        error: err.response?.data?.detail || err.message
      }
    }
  })

  // Add a node to a project
  ipcMain.handle('project:add-node', async (_event, data) => {
    try {
      const { projectId, node } = data
      const response = await axios.post(
        `${baseUrl}/projects/${projectId}/nodes`,
        node
      )
      return { success: true, data: response.data }
    } catch (err) {
      return {
        success: false,
        error: err.response?.data?.detail || err.message
      }
    }
  })

  // Update a node in a project
  ipcMain.handle('project:update-node', async (_event, data) => {
    try {
      const { projectId, nodeId, updates } = data
      const response = await axios.put(
        `${baseUrl}/projects/${projectId}/nodes/${nodeId}`,
        updates
      )
      return { success: true, data: response.data }
    } catch (err) {
      return {
        success: false,
        error: err.response?.data?.detail || err.message
      }
    }
  })

  // Remove a node from a project
  ipcMain.handle('project:remove-node', async (_event, data) => {
    try {
      const { projectId, nodeId } = data
      const response = await axios.delete(
        `${baseUrl}/projects/${projectId}/nodes/${nodeId}`
      )
      return { success: true, data: response.data }
    } catch (err) {
      return {
        success: false,
        error: err.response?.data?.detail || err.message
      }
    }
  })
}

module.exports = { register }
